Make libtorch buildable with buck1 & buck2

arlyon opened the pull request:
This change makes a few changes to the buck setup script, and to the build definitions to allow building the project with buck1 _or_ two. Because of the prelude, the buckconfig files are incompatible so we instead have two. Note that this is building with the oss-ready buck2 build which is not currently freely available but we expect it to be ready before the end of the year. 

To make this work, we update some imports, and have a minimal prelude shim which buck1 can use to expose the appropriate methods in the correct places. These imports are bundled in the buck2 prelude.

To set up the project using buck1:

```
git clone --recursive https://github.com/pytorch/pytorch.git
cd pytorch
bash ./scripts/buck_setup.sh
buck build :torch
```

To set up the project for buck2:

```
git clone --recursive https://github.com/pytorch/pytorch.git
cd pytorch
bash ./scripts/buck_setup.sh 2
buck2 build :torch
```

The buck2 setup command uses `buck2 init`, so you will need a copy of buck2 to run it. It also relies on a few patches that still need to land in the prelude, but this will happen before buck2 oss is released anyways. Note that, until the prelude repo is made public, the buck2 setup script will fail, since it needs to clone the prelude as a submodule.

The next step is to get pytorch bindings building with buck2. I have a patch in progress for this but it is not yet finished. It does not aim to also be compatible with buck1 (however it is reasonable to expect we could make it compatible with some tweaks)

Some notes:

- I was getting build errors on a recent version of clang, because `__has_trivial_copy` has gone from a deprecation warning to a hard error. I naively followed the suggestion of the warning and put the code change in its own commit.
- buck2 doesn't (currently) support the .buckconfig `includes` directive in `.bzl` files. this is going to be fixed at some point but in the mean time we just import the 'patch_select.bzl' file to fix those cases.
- prelude-shim is included by default in the root. we could perhaps move it if needed
- the 'devserver' param for the buck2_setup command has been moved to the second parameter, and is only respected when generating for buck1
- a simple toolchains folder has been added. we are currently in the process of moving some basic system toolchains into the prelude however, so `toolchains/toolchain.bzl` will be much smaller. I originally looked at using our hermetic cpp toolchain, but it does not currently support asm, so the toolchain is just using whatever is on your system

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel @VitalyFedyunin @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

arlyon committed:
replace use of deprecated __has_trivial_copy

arlyon committed:
patch `select` for buck2

Currently buck2 does not handle includes in bzl files. This will be
fixed upstream, and the load directives can be removed.

arlyon committed:
adjust imports for buck2

arlyon committed:
add .buckconfig.1.oss and .buckconfig.2.oss

arlyon committed:
fixups for compat with buck2

arlyon committed:
add toolchains for buck2

arlyon committed:
add prelude-shim for buck 1

arlyon committed:
add buck2 to bootstrap script

ndmitchell commented on tools/build_defs/fb_xplat_genrule.bzl:
```diff
@@ -1,6 +1,8 @@
 # Only used for PyTorch open source BUCK build
 # @lint-ignore-every BUCKRESTRICTEDSYNTAX
 
+load("@prelude//:rules.bzl", "genrule")
```
Rather than do this import, you should do `native.genrule` further down, which should work on both Buck1 and Buck2.

ndmitchell commented on tools/build_defs/fb_xplat_cxx_test.bzl:
```diff
@@ -1,6 +1,7 @@
 # Only used for PyTorch open source BUCK build
 # @lint-ignore-every BUCKRESTRICTEDSYNTAX
 load(":buck_helpers.bzl", "filter_attributes")
+load("@prelude//:rules.bzl", "cxx_test")
```
Ditto, `native.cxx_test`

ndmitchell commented on tools/build_defs/fb_xplat_cxx_library.bzl:
```diff
@@ -1,6 +1,8 @@
 # Only used for PyTorch open source BUCK build
 # @lint-ignore-every BUCKRESTRICTEDSYNTAX
 
+load("@prelude//:rules.bzl", "cxx_library")
```
Ditto

ndmitchell commented on tools/build_defs/fb_native_wrapper.bzl:
```diff
@@ -1,3 +1,5 @@
+load("@prelude//:rules.bzl", "genrule", "filegroup")
```
Ditto

ndmitchell commented on toolchains/.buckconfig:
```diff
@@ -0,0 +1,2 @@
+[buildfile]
+name=BUILD
```
Why not BUCK.oss given this is what is used elsewhere?

ndmitchell commented on third_party/kineto.buck.bzl:
```diff
@@ -1,4 +1,5 @@
 load("//tools/build_defs:glob_defs.bzl", "subdir_glob")
+load("@prelude//:rules.bzl", "cxx_library", "genrule")
```
Ditto, use native

ndmitchell commented on scripts/buck_setup.sh:
```diff
@@ -1,17 +1,31 @@
 #!/bin/bash
 printf "\nCreating .buckconfig\n"
```
Would be much better if this used `set -e`, so errors were obvious

arlyon replied:
Agreed

ndmitchell commented on prelude-shim/rules.bzl:
```diff
@@ -0,0 +1,4 @@
+genrule = native.genrule
```
I think this can be fixed by using `native.genrule` in the right location, as that will work in Buck1 and Buck2. Then you can delete prelude-shim

ndmitchell commented on c10/util/variant.h:
```diff
@@ -878,8 +878,8 @@ using std::is_trivially_move_constructible;
-    : bool_constant<std::is_copy_constructible<T>::value&& __has_trivial_copy(
-          T)> {};
+    : bool_constant<
+          std::is_copy_constructible<T>::value&& __is_trivially_copyable(T)> {};
```
Why is this changing?

arlyon replied:
`__has_trivial_copy` is deprecated (I think for a decade) and as of clang version 15 is a build error on my machine. `__is_trivially_copyable` is recommended by the compiler, and supported by both clang and gcc.

ndmitchell commented on c10/macros/Macros.h:
```diff
@@ -431,11 +431,11 @@ __device__ __attribute__((noinline)) __attribute__((weak)) void __assert_fail(
 
 // Portable determination of whether type T is trivially copyable.
-// Warning: __has_trivial_copy for GCC may not always detect the non-POD
+// Warning: __is_trivially_copyable for GCC may not always detect the non-POD
```
Why is this changing?

arlyon replied:
I left a comment here, not sure if we still need this warning as __is_trivially_copyable is much 'better'. I updated the comment just in case but I assume we can get rid of this.

ndmitchell commented on toolchains/toolchain.bzl:
```diff
@@ -0,0 +1,154 @@
+# This source code is licensed under both the MIT license found in the
+# LICENSE-MIT file in the root directory of this source tree and the Apache
+# License, Version 2.0 found in the LICENSE-APACHE file in the root directory
+# of this source tree.
```
I've just put up D42040624 to move the toolchains into the prelude, so they can be reused here rather than copied (one of them has broken in the last two weeks since you released this)

arlyon committed:
address PR comments

- prefer native.X rather than a stub cell
- add set -e to the setup script
- remove all prelude imports from the build files
- remove custom toolchains in favour of prelude
- make toolchain cell use BUCK.oss build file

arlyon commented:
I am still waiting for review.

arlyon commented:
Hi, this PR was automatically closed. Is there a reason why? What can I do to help push this forward?
